fastcrypto-tbls/src/weight_reduction/mod.rs

```diff
@@ -0,0 +1,552 @@
+            deltas_left,
+        ) {
+            if process_batch_recursive(beta, weights, deltas_left, &dp, tickets) {
+                return true;
+            }
+        } else {
+            // Apply the left deltas before continuing.
+            tickets.update_many(deltas_left);
+        }
+        tickets.update(deltas_apply[0]);
+    }
+
+    {
+        let (deltas_apply, deltas_right) = deltas.split_at(left_branch_size);
+        if let Some(dp) = apply(
+            weights,
+            dp_head,
+            tickets,
+            calc_adv_tickets_target(beta, tickets.total() + deltas_right.len() as u64),
+            deltas_apply,
+            deltas_right,
+        ) {
+            process_batch_recursive(beta, weights, deltas_right, &dp, tickets)
+        } else {
+            // Apply the rest of the deltas before exiting.
+            tickets.update_many(deltas_right);
+            false
+        }
+    }
+}
+
+/// Apply `deltas` to provided `tickets`. If after applying 0 or more
+/// deltas a valid ticket assignment is found, returns true with
+/// `tickets` containing the corresponding ticket assignment.
+/// Otherwise, returns false with `tickets` containing
+/// the new ticket assignment after applying all deltas.
+fn process_batch(
+    beta: Ratio,
+    weights: &[u64],
+    max_adv_weight: u64,
+    deltas: &[usize],
+    tickets: &mut Tickets,
+) -> bool {
+    let Some(dp_head) = calc_dp_head(beta, weights, max_adv_weight, deltas, tickets) else {
+        // We are exiting early. Apply all of the deltas before that.
+        tickets.update_many(deltas);
+        return false;
+    };
+
+    process_batch_recursive(beta, weights, deltas, &dp_head, tickets)
+}
+
+pub fn solve(alpha: Ratio, beta: Ratio, weights: &[u64]) -> Vec<u64> {
+    debug_assert!(weights.windows(2).all(|w| w[0] >= w[1]));
+
+    let max_adv_weight = calc_max_adv_weight_from_weights(alpha, weights);
+
+    let mut tickets = Tickets::new();
+    let mut g = generate_deltas(weights, alpha);
+
+    let mut batch_size: usize = 1;
+    // This loop terminates because:
+    // 1. The generator `g` is infinite (always produces deltas via the queue-based algorithm)
+    // 2. We use exponential backoff: batch_size doubles each iteration (1, 2, 4, 8, ...)
+    // 3. The algorithm is guaranteed to find a solution (by the paper's theoretical results)
+    // 4. Once batch_size is large enough to include all deltas needed for a valid solution,
+    //    `process_batch_recursive` will eventually return true when it successfully applies
+    //    all deltas in the batch without the adversary winning (see the empty deltas base case)
+    loop {
+        tickets.update(g.next().unwrap());
+        let deltas: Vec<_> = (&mut g).take(batch_size - 1).collect();
+
+        let ret = process_batch(beta, weights, max_adv_weight, &deltas, &mut tickets);
+        if ret {
+            return tickets.into_vec();
+        }
+
+        batch_size *= 2;
```

Comment:
Is there an upper bound on the number of iterations which guarantees that this doesn't overflow?